File: LinkedList/LinkedList.cpp
#include <iostream>

class Node
{
public:
  int value;
  class Node *next;
  Node(int v) : value(v), next(nullptr) {};
};

class LinkedList
{
public:
  LinkedList();
  ~LinkedList();
  int size();
  void append(int v);
  bool remove(int v);
  void traverse();
  int getTail();

private:
  Node *head;
  Node *tail;
  int count;
};

LinkedList::LinkedList()
: head(nullptr),
  tail(nullptr),
  count(0)
{}

LinkedList::~LinkedList()
{
  Node *n = head;
  while (n != nullptr)
  {
    Node *next = n->next;
    delete(n);
    n = next;
    count--;
  }
}

int LinkedList::size()
{
  return LinkedList::count;
}

void LinkedList::append(int v)
{
  Node *n = new Node(v);
  if (head == nullptr)
  {
    head = n;
    tail = head;
  }
  else
  {
    tail->next = n;
    tail = n;
  }
  count++;
}

bool LinkedList::remove(int v)
{
  Node *n = head;
  Node *prev = nullptr;
  
  if (head == nullptr)
    return false;
  
  while(n != nullptr)
  {
    std::cout << "Node: " << n->value << std::endl;
    if (n->value == v)
    {
      if ((n == head) && (n == tail))
      {
        std::cout << "Head & Tail" << std::endl;
        delete(n);
        head = nullptr;
        tail = nullptr;
        count--;
        return true;
      }
      
      if (n == head)
      {
        std::cout << "Head Only" << std::endl;
        head = n->next;
        prev = n;
        delete(n);
        count--;
        return true;
      }
      else
      {
        std::cout << "Middle" << std::endl;
        prev->next = n->next;
        delete(n);
        count--;
        return true;
      }
    }
    else
    {
      prev = n;
      n = n->next;
    }
  }
  return false;
}

void LinkedList::traverse()
{
  Node *n = head;
  while (n != nullptr)
  {
    std::cout << n->value << " ";
    n = n->next;
  }
  std::cout << std::endl;
}

int LinkedList::getTail()
{
  return tail->value;
}

/* Test */
int main()
{
  int a[] = {1,2,3,4,5,6,7,8,9,10};
  LinkedList l;
  for (int i = 0; i < 10; i++)
  {
    l.append(a[i]);
  }

  l.traverse();
  std::cout << "Size: " << l.size() << std::endl;

  l.remove(1);
  l.traverse();

  l.remove(2);
  l.traverse();

  l.remove(10);
  l.traverse();

  l.remove(5);
  l.traverse();
  std::cout << "Size: " << l.size() << std::endl;

  return 0;
}

File: LinkedList/LinkedList.py
#!/usr/bin/env python
# 2021-03-25: ST

# Represents a single element. It contains 
class Element:

    # Cpnstructor for an element
    def __init__(self, val):
        self.value = val                                       # the value of the element
        self.next = None                                       # Next Element


# A class to control the entire list
class LinkedList:

    # Create an empty list
    def __init__(self):
        self.head = None
        self.count = 0                                         # Just for convinience


    # Return the number of elements
    def size(self):
        return self.count


    # Convert the Linked list to an array
    def toList(self):
        if self.head == None:                                  # If the list is empty,
            return []                                          # Return empty list

        ret = []
        current_element = self.head                            # Starting from head
        while True:
            ret.append(current_element.value)
            self.count = self.count + 1
            if current_element.next == None:                   # End of the list
                break
            current_element = current_element.next

        return ret


    # Append an element to the end.
    def append(self, val):
        if self.head == None:                                  # The first element
            self.head = Element(val)
            self.count = 1
            return

        last_element = self.head                               # Starting from head
        while last_element.next != None:                       # Traverse to the end.
            last_element = last_element.next

        last_element.next = Element(val)
        self.count = self.count + 1


    # Delete the first element that has the value specified
    # Returns False if the element is not found. Otherwise True.
    def delete(self, val):
        if self.head == None:
            return False

        if self.head.next == None and self.head.value == val:  # It has only one element and that's what I need to delete
            self.head = None
            self.count = 0
            return True

        prev_element = None
        current_element = self.head
 
        while current_element.next != None:
            if current_element.value == val:
                if current_element.next == None:               # This is the last element. Simply remove the previous next
                    prev_element.next = None
                    self.count = self.count - 1
                    print('Got ', val)
                    return True
                else:                                          # The element has the next one
                    prev_element.next = current_element.next
                    self.count = self.count - 1
                    print('Got ', val)
                    return True
            prev_element = current_element
            current_element = current_element.next

        return False


### Test
if __name__ == '__main__':
    linkedList = LinkedList()
    for i in range(0, 20, 2):
        linkedList.append(i)
    print('%d: %s' % (linkedList.size(), linkedList.toList()))

    linkedList.delete(10)
    print('%d: %s' % (linkedList.size(), linkedList.toList()))


